src/linker/Linker/UnconditionalSuppressMessageAttributeState.cs

```diff
@@ -99,10 +106,127 @@ private static bool TryDecodeSuppressMessageAttributeData (CustomAttribute attri
 
 		private bool IsLocallySuppressed (int id, ICustomAttributeProvider provider, out SuppressMessageInfo info)
 		{
-			Dictionary<int, SuppressMessageInfo> suppressions;
 			info = default;
-			return _localSuppressions.TryGetValue (provider, out suppressions) &&
+			if (provider == null)
+				return false;
+
+			return _localSuppressions.TryGetValue (provider, out var suppressions) &&
 				suppressions.TryGetValue (id, out info);
 		}
+
+		private bool IsGloballySuppressed (int id, ICustomAttributeProvider provider, out SuppressMessageInfo info)
+		{
+			DecodeGlobalSuppressMessageAttributes (provider);
+			return _globalSuppressions.HasGlobalSuppressions (id, provider, out info);
+		}
+
+		private void DecodeGlobalSuppressMessageAttributes (ICustomAttributeProvider provider)
+		{
+			ModuleDefinition module;
+			switch (provider.MetadataToken.TokenType) {
+			case TokenType.Module:
+				module = provider as ModuleDefinition;
+				break;
+			case TokenType.Assembly:
+				module = (provider as AssemblyDefinition).MainModule;
+				break;
+			case TokenType.TypeDef:
+				module = (provider as TypeDefinition).Module;
+				break;
+			case TokenType.Method:
+			case TokenType.Property:
+			case TokenType.Field:
+			case TokenType.Event:
+				module = (provider as IMemberDefinition).DeclaringType.Module;
+				break;
+			default:
+				_context.LogMessage (MessageContainer.CreateInfoMessage ("`UnconditionalSuppressMessage` attribute was placed in an language element which is currently not supported."));
+				return;
+			}
+
+			Debug.Assert (module != null);
+			AssemblyDefinition assembly = module.Assembly;
+			if (!_globalSuppressions.InitializedAssemblies.Contains (assembly)) {
+				DecodeGlobalSuppressMessageAttributes (assembly);
+				_globalSuppressions.InitializedAssemblies.Add (assembly);
+			}
+		}
+
+		private void DecodeGlobalSuppressMessageAttributes (AssemblyDefinition assembly)
+		{
+			void LookForSuppressions (ICustomAttributeProvider provider) {
+				ModuleDefinition module = provider is ModuleDefinition ? provider as ModuleDefinition : (provider as AssemblyDefinition).MainModule; 
+				var attributes = provider.CustomAttributes.
+					Where (a => a.AttributeType.Name == "UnconditionalSuppressMessageAttribute" && a.AttributeType.Namespace == "System.Diagnostics.CodeAnalysis");
+				foreach (var instance in attributes) {
+					SuppressMessageInfo info;
+					if (!TryDecodeSuppressMessageAttributeData (instance, out info))
+						continue;
+
+					if (info.Target == null && (info.Scope == "module" || info.Scope == null)) {
+						AddLocalSuppression (instance, provider);
```

Comment:
Naming is weird here. The method we're in is called `DecodeGlobal...` but here we're adding local suppressions.

Reply:
I guess I find it weird how we store the suppressions.
Maybe a better design would be to have one storage (the one which is currently called local) and then each time check against it we first make sure that global suppressions where processed (initialized), those would all feed into the same storage at the end.

If you look at the two local/global, they store the exact same information in the end - same dictionary, same kind of lookup and so on. The only difference is how they get initialized.